src/Transposit.ts

```diff
@@ -15,144 +15,120 @@
  */
 
 import { EndRequestLog } from ".";
-
-export interface ClientClaims {
-  iss: string; // issuer
-  sub: string; // subject
-  exp: number; // expiration
-  iat: number; // issuedAt
-  publicToken: string;
-  repository: string;
-  email: string;
-  name: string;
-}
-
-export interface OperationParameters {
-  [paramName: string]: string;
-}
-
-// https://my-app.transposit.io?clientJwt=...needKeys=... -> https://my-app.transposit.io
-function hereWithoutSearch(): string {
-  return `${window.location.origin}${window.location.pathname}`;
-}
-
-export const LOCAL_STORAGE_KEY = "TRANSPOSIT_SESSION";
+import { popCodeVerifier, pushCodeVerifier } from "./pkce-helper";
+import {
+  clearPersistedData,
+  loadAccessToken,
+  loadUser,
+  persistAccessToken,
+  persistUser,
+  TokenResponse,
+  User,
+} from "./token";
+import { chompSlash, formUrlEncode, hereWithoutSearch } from "./util";
 
 export class Transposit {
-  private claims: ClientClaims | null = null;
+  private hostedAppOrigin: string;
+  private accessToken: string | null;
+  private user: User | null;
+
+  constructor(hostedAppOrigin: string = "") {
+    this.hostedAppOrigin = chompSlash(hostedAppOrigin);
+    this.accessToken = null;
+    this.user = null;
 
-  constructor(private hostedAppOrigin: string = "") {
-    this.claims = this.loadClaims();
+    this.load();
   }
 
   private uri(path: string = ""): string {
     return `${this.hostedAppOrigin}${path}`;
   }
 
-  private loadClaims(): ClientClaims | null {
-    const claimsJSON = localStorage.getItem(LOCAL_STORAGE_KEY);
-    if (!claimsJSON) {
-      return null;
-    }
-
-    let claims: ClientClaims;
-    try {
-      claims = JSON.parse(claimsJSON);
-    } catch (err) {
-      return null;
-    }
-
-    if (!this.checkClaimsValid(claims)) {
-      return null;
-    }
-
-    return claims;
+  private load(): void {
+    this.accessToken = loadAccessToken();
+    this.user = loadUser();
   }
 
-  private persistClaims(claimsJSON: string): void {
-    localStorage.setItem(LOCAL_STORAGE_KEY, claimsJSON);
+  private assertSignedIn(): void {
+    if (this.accessToken === null || this.user === null) {
+      throw new Error("Missing accessToken or user. No one is signed in.");
+    }
   }
 
-  private clearClaims(): void {
-    localStorage.removeItem(LOCAL_STORAGE_KEY);
+  isSignedIn(): boolean {
+    return this.accessToken !== null && this.user !== null;
   }
 
-  private checkClaimsValid(claims: ClientClaims): boolean {
-    const expiration = claims.exp * 1000;
-    const now = Date.now();
-    return expiration > now;
-  }
+  async signIn(
+    redirectUri: string,
+    provider?: "google" | "slack",
+  ): Promise<void> {
+    const codeChallenge = await pushCodeVerifier();
 
-  private assertLoggedIn(): void {
-    if (this.claims === null) {
-      throw new Error("No client claims found.");
+    const params = new URLSearchParams();
+    params.append("scope", "openid");
```

Comment:
In that case you'll need handle different scopes in the server, right? (both in the code that generate the token and in the endpoints auth filter)
I assume we don't do that already, and in that case I agree to keep this as a future improvement (maybe together with role based tokens).

Reply:
It's a good idea for sake of completion because accessToken isn't technically part of openid, but I agree with Morad. It would be fine as future work

Reply:
I'm going to add "app" as a scope. This will be meaningless, but if we one day want to assign meaning to it, we can.